### PR TITLE
doc: input: add a dedicated GPIO keyboard documentation

### PR DESCRIPTION
**Pull request by fabiobaltieri**
This driver supports few different hardware configurations, add a document page showing some of them next to some schematics snippets

https://builds.zephyrproject.io/zephyr/pr/66450/docs/services/input/gpio-kbd.html

### CODE REVIEW

**Review by gmarull — APPROVED** [review]
nice circuit diagrams

**Review by aasinclair — APPROVED** [review]

File: doc/services/input/gpio-kbd.rst [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+automatically switches API to read from the whole GPIO port rather than the
+individual pins. This is particularly useful if the GPIOs are not memory
+mapped, for example on an I2C or SPI port expander, as this significantly
+reduces the number of transactions on the corresponding bus.
+
+The same is true for column GPIOs, but only if the matrix is configured for
+``col-drive-inactive``, so that is only usable for matrixes with isolation
+diodes.
+
+16 bit row support
```
Comment (kartben) [inline doc/services/input/gpio-kbd.rst:163]:
```suggestion
16-bit row support
```

File: doc/services/input/gpio-kbd.rst [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+******************
+
+The driver uses an 8 bit datatype to store the row state by default, which
```
Comment (kartben) [inline doc/services/input/gpio-kbd.rst:166]:
```suggestion
The driver uses an 8-bit datatype to store the row state by default, which
```

File: doc/services/input/gpio-kbd.rst [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+.. code-block:: devicetree
+
+   kbd-matrix {
+        compatible = "gpio-kbd-matrix";
+        ...
+        actual-key-mask = <0x07 0x05 0x07>;
+   };
+
+Would allow, for example, to detect pressing ``Sw0``, ``SW1`` and  ``SW3`` at
```
Comment (kartben) [inline doc/services/input/gpio-kbd.rst:195]:
```suggestion
This would allow, for example, to detect pressing ``Sw0``, ``SW1`` and  ``SW3`` at
```

Comment (kartben) [reply doc/services/input/gpio-kbd.rst:195]: also, `SW0` does not exist (maybe you used different numbering scheme before), and I am terrible at ghosting so can't suggest the proper change :)

File: doc/services/input/gpio-kbd.rst [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+the same time without triggering anti ghosting.
+
+The actual key mask can be changed in runtime by enabling
```
Comment (kartben) [inline doc/services/input/gpio-kbd.rst:198]:
```suggestion
The actual key mask can be changed at runtime by enabling
```

**Review by kartben — COMMENTED** [review]
Looks like i was too slow to hit submit this morning :/
The requested changes are mostly minor spelling/grammar issues but there is also an actual error in the text accompanying the ghosting diagram

**Comment by kartben** [comment]
BTW: this is a really nice doc page! <3

File: doc/services/input/gpio-kbd.rst (hunk 3 above)
Comment (fabiobaltieri) [reply doc/services/input/gpio-kbd.rst:195]: Ah! yeah I originally had the illustration refdes numbered from 0, then I moved them around in the CAD and the thing automatically renumbered them (it tries REALLY hard to keep them unique), so I had to rename it again and I guess at that point I had my wannabe-EE hat on an started counting from 1, which is the right thing to do in this context I think, I'll rename them to 1, 2, 4.

I actually wanted to do an animation of the whole scanning story, looked a bit for online tools that could export something embeddable here but did not find any, too bad :-), the ghosting concept is very simple, I may add some extra illustration down the road.

**Comment by fabiobaltieri** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/67308 for the followups.

**Pull request merged**